File: src/Block/Block.java
/**
 * This file copyright (c) 2015-2016, Jing Fan, Adalbert Gerald Soosai Raj, and Jignesh M. Patel
 *
 * See the file CREDITS.txt in the root directory for details.
 **/

package Block;

/**
 * @brief Basic SQL statement block.
 */
public class Block {
  // Description of the target of this block. Such as "sendMsg", "combineMsg".
  protected String stage;
  // Buffer for intermediate SQL.
  protected StringBuilder sb;
  // SQL statement.
  protected String sql;
  // The initial level of indentation of this block.
  protected int indentLevel;

  /**
   * @brief Constructor.
   * @param stage The stage.
   * @param indentLevel The level of indentation.
   * @param sql The sql statement.
   */
  public Block(String stage, int indentLevel, String sql) {
    this.stage = stage;
    this.indentLevel = indentLevel;
    this.sb = new StringBuilder();
    this.append(sql);
    this.sql = this.sb.toString();
  }

  /**
   * @brief Constructor.
   * @param stage The stage.
   * @param sql The sql statement.
   */
  public Block(String stage, String sql) {
    this.stage = stage;
    this.sql = sql;
  }

  /**
   * @brief Constructor.
   * @param stage The stage.
   * @param indentLevel The level of indentation.
   */
  public Block(String stage, int indentLevel) {
    this.stage = stage;
    this.indentLevel = indentLevel;
    this.sb = new StringBuilder();
  }

  /**
   * @brief Getters and Setters.
   */
  public String getStage() {
    return this.stage;
  }

  public String getSql() {
    this.sql = this.sb.toString();
    return this.sql;
  }

  public int getIndentLevel() {
    return indentLevel;
  }

  public void setStage(String stage) {
    this.stage = stage;
  }

  public void setSql(String sql) {
    this.sql = sql;
  }

  public void setIndentLevel(int indentLevel) {
    this.indentLevel = indentLevel;
  }

  /**
   * @brief Print the information of the block.
   */
  public void print() {
    if (sql == null) {
      sql = this.sb.toString();
    }
    System.out.println(sql);
  }

  /**
   * @brief Append SQL with indentation.
   * @param stat The SQL statement.
   * @param indent Indentation level.
   */
  public void append(String stat, int indent) {
    StringBuilder line = new StringBuilder();
    for (int i = 0; i < indent; ++i) {
      line.append("  ");
    }
    line.append(stat);
    this.sb.append(line);
    this.sb.append("\n");
  }

  public void concat(String stat) {
    this.sb.append(stat);
  }

  public void append(String stat) {
    this.append(stat, this.indentLevel);
  }

  public String toString() {
    return this.sb.toString();
  }
}

File: python/src/Blocks/Block.py
class Block(object):
    sb=""
    sql=""
    indentLevel=-1
    def __init__(self,stage="",indentLevel=-1,sql=""):
        '''
        constructor for the class; generates the sql statements for a generic block
        @param stage:stage of the code block
        @type stage:string
        @param indentLevel: indent level of the statements
        @type indentLevel: int
        @param sql: sql statements in the blocl
        @type sql: string
        '''
        self.stage=stage
        self.indentLevel=indentLevel
        self.sb=""
        if(sql!="" and indentLevel!=-1):
            self.append(sql)
            self.sql=self.sb
        elif(sql!="" and indentLevel==-1):
            self.sql=sql
    def concat(self,stat):
        '''
        concatenates string to the existing sql statement
        @param stat: string to concat
        @type stat: string
        '''
        self.sb+=stat
    def append(self,stat,indent=-1):
        '''
        appends string to the existing sql 
        @param stat: string to append
        @type stat: string
        @param indent: indent level current
        @type indent: int
        '''
        if(indent==-1):
            self.append(stat,self.indentLevel)
        else:
            line=""
            for i in range(0,indent):
                line+=" "
            line+=stat
            self.sb+=line 
            self.sb+="\n"           
    def getStage(self):
        '''
        returs the stage
        '''
        return self.stage
    def getSql(self):
        '''
        returns the sql statements
        '''
        self.sql=self.sb
        return self.sql
    def getIndentLevel(self):
        '''
        returns the indent level
        '''
        return self.indentLevel
    def setStage(self,stage):
        '''
        sets the stage
        @param stage: string containing the stage of the block
        @type stage: string
        '''
        self.stage=stage
    def setSql(self,sql):
        '''
        sets the sql into the block
        @param sql: string containing sql
        @type sql: string
        '''
        self.sql=sql
    def setIndentLevel(self,indent):
        '''
        sets the indent level
        @param indent: indent level
        @type indent: int
        '''
        self.indentLevel=indent
    def display(self):
        '''
        displays the sql statements
        '''
        if(self.sql==""):
            self.sql=self.sb
        print self.sql                    